Stripe/STPSourcePoller.m

```diff
@@ -0,0 +1,170 @@
+                                   name:UIApplicationWillEnterForegroundNotification
+                                 object:nil];
+        [notificationCenter addObserver:self
+                               selector:@selector(stopPolling)
+                                   name:UIApplicationWillResignActiveNotification
+                                 object:nil];
+        [notificationCenter addObserver:self
+                               selector:@selector(stopPolling)
+                                   name:UIApplicationDidEnterBackgroundNotification
+                                 object:nil];
+    }
+    return self;
+}
+
+- (void)dealloc {
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)pollAfter:(NSTimeInterval)interval {
+    self.timer = [NSTimer scheduledTimerWithTimeInterval:interval
+                                                  target:self
+                                                selector:@selector(poll)
+                                                userInfo:nil
+                                                 repeats:NO];
+}
+
+- (void)poll {
+    NSTimeInterval totalTime = [[NSDate date] timeIntervalSinceDate:self.startTime];
+    if (!self.apiClient || totalTime >= Timeout || self.retryCount >= MaxRetries) {
+        [self stopPolling];
+        return;
+    }
+    self.dataTask = [self.apiClient retrieveSourceWithId:self.sourceID
+                                            clientSecret:self.clientSecret
+                                      responseCompletion:^(STPSource *source, NSHTTPURLResponse *response, NSError *error) {
+                                          [self continueWithSource:source response:response error:error];
+                                      }];
+}
+
+- (void)restartPolling {
+    if (!self.timer && !self.dataTask) {
+        [self pollAfter:0];
+    }
+}
+
+- (void)continueWithSource:(STPSource *)source
+                  response:(NSHTTPURLResponse *)response
+                     error:(NSError *)error {
+    if (response) {
+        NSUInteger status = response.statusCode;
+        if (status >= 400 && status < 500) {
+            // Don't retry requests that 4xx
+            self.completion(self.latestSource, error);
+            [self stopPolling];
+        } else if (status == 200) {
+            self.pollInterval = DefaultPollInterval;
+            // Only call completion if source.status has changed
+            if (!self.latestSource || source.status != self.latestSource.status) {
+                self.completion(source, nil);
+            }
+            self.latestSource = source;
+            if ([self shouldContinuePollingSource:source]) {
+                [self pollAfter:self.pollInterval];
+            } else {
+                [self stopPolling];
+            }
+            self.retryCount = 0;
+        } else {
+            // Backoff on 500, otherwise reset poll interval
+            if (status == 500) {
```

Comment:
I think backoff is only necessary if there's a server error – 5xx would probably be better. Do you think we should backoff in other cases too?